experimental/dds/tree/src/HistoryEditFactory.ts

```diff
@@ -72,7 +75,6 @@ export function revert(changes: readonly ChangeInternal[], before: RevisionView)
 				const { source } = change;
 				const nodesBuilt = builtNodes.get(source);
 				const nodesDetached = detachedNodes.get(source);
-
 				if (nodesBuilt !== undefined) {
```

Comment:
We should do a continue here as well just in case we are missing any other scenarios where empty inserts could be added. Continuing in the insert case is necessary because it's a fatal error otherwise. It's not necessary in the detach case but reduces edit bloat.

Reply:
I'm also surprised we're not doing any logging in here, I guess this code predated out logger. I think it would be a good idea to log if we ever do run into empty inserts so we're aware something weird is happening.

Reply:
Oh never mind, nodesBuilt would be undefined in every case and I guess you haven't made the changes from the conversation with Noah yet.

Logging in the build case would be good then.

Reply:
Perfect, thank you for the comments :)

I have now added the changes to now skip the change for the insert/detach if there is an empty trait. I have also updated the testcases (as they no longer return undefined).

I am not sure if these new test cases are sufficient, but for now I have added the following 2:
1. build and insert an empty traits, followed by changes with multiple inserts with non-empty traits. The expected result is that the first build and insert is ignored, and the rest of the changes were applied.
2. detaching an empty trait (or a trait that doesn't exist in the tree), where the expected result is to be length 0, as the change would be skipped.

Please let me know if I should fix/cover any additional test cases

Reply:
@jenn-le Are you sure we want to return undefined? It seems to me that there's a distinction here that could matter. For example, suppose we've hookup up some UI to the revert function and if it returns undefined, we show a "Failed to undo!" toast popup to the user. We wouldn't want that in this case; we'd want the undo to "succeed", but do nothing.

I do think we should at least document that the revert function can sometimes return an empty array.

Reply:
I think it's a question of how we're currently treating undefined right now then. It looks like we don't do anything if a revert returns undefined but otherwise, we just try to apply the change: https://github.com/microsoft/FluidFramework/blob/96b92b04158fdae99eda7a5a8628ec106e45282a/experimental/dds/tree/src/Checkout.ts#L371

If we want it to be possible for the app author to deal with the different failure scenarios however they choose, then I think we should be emitting events.

Reply:
That's a good idea - we could also do that, especially if we want telemetry about why a failure occurs in the future. But as it stands, this current change communicates useful semantics and is totally compatible with the existing Checkout code. Checkout will apply the edit that has 0 changes, and do nothing, which is fine.

Reply:
Does checkout send anything over the wire if there are no changes? If not, either case is fine. I like undefined more for consistency but either way, we should document the difference. I don't think undefined vs empty array communicates much to someone new to the code.

Reply:
Tldr; I think you are right that we should skip processing empty reverts. I think that it is Checkout's responsibility, or EditLog's responsibility, not revert's.

Last I checked, Checkout does not filter out empty edits, nor does tree. I actually tried to do that a while ago but a bunch of Whiteboard document code had taken a dependency on the existing behavior for some reason (rip). Perhaps it could still be done though, if we want to do that. 

But regardless, my view of the current behavior is the following. Revert can return two things. Either:
1. The edit passed to revert was revertible, so it returns the reverted edit.
2. The edit passed to revert was not revertible, so it returns undefined.
By "revertible" here I mean that it was able to make an edit that returns the tree state to what it was before the original edit, given the tree state after the original edit.

Daesun's change here is introducing additional possibilities in case 1, _not case 2_. The scenarios he's handling are both valid and invertible. The minimal edit required to revert them happens to be an empty edit. I think that the revert API should return information about its success rather than returning an instruction that the caller is forced to do; i.e. `undefined` should mean "I could not revert this edit" rather than "I'm deciding that you should skip this edit". This is purer - revert only thinks about the reversion itself, it does not think about how to apply the reversion. The caller gets to decide how to apply it (e.g. skip empty edits. Or, maybe not! What if it wants to count empty edits for telemetry or something? etc.).

Reply:
That's a good point. Especially since revert semantically operates on a group of changes rather than an edit as a whole. I'm convinced on returning an empty array but still think the difference should be documented, it's not a super obvious distinction.

Reply:
Completely agree. What do you think about adding something like "Note that the array of reverted changes may not be the same size as the input array, and may even be empty in cases where the view did not change" between these sentences:
<img width="725" alt="image" src="https://user-images.githubusercontent.com/78610362/182438715-4395c0a5-c7a5-4807-a3fb-d2c458881f8f.png">

Reply:
Sounds good!